builtin/nomad/jobspec/releaser.go

```diff
@@ -0,0 +1,547 @@
+			resource.WithStatus(r.resourceJobStatus),
+			resource.WithPlatform("nomad"),
+			resource.WithCategoryDisplayHint(sdk.ResourceCategoryDisplayHint_INSTANCE_MANAGER),
+		)),
+	)
+}
+
+// getNomadClient provides
+// the client connection used by resources to interact with Nomad.
+func (r *Releaser) getNomadClient() (*nomadClient, error) {
+	// Get our client
+	client, err := api.NewClient(api.DefaultConfig())
+	if err != nil {
+		return nil, err
+	}
+	return &nomadClient{
+		NomadClient: client,
+	}, nil
+}
+
+func (r *Releaser) resourceJobCreate(
+	ctx context.Context,
+	log hclog.Logger,
+	target *Deployment,
+	result *Release,
+	state *Resource_Job,
+	client *nomadClient,
+	st terminal.Status,
+) error {
+	// Set up clients
+	jobClient := client.NomadClient.Jobs()
+	deploymentClient := client.NomadClient.Deployments()
+
+	st.Update("Getting job...")
+	jobs, _, err := jobClient.PrefixList(target.Name)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad jobs: %s", err.Error())
+	} else if target.Name != jobs[0].ID {
+		return status.Errorf(codes.Aborted, "Job not found: %s", err.Error())
+	}
+
+	q := &api.QueryOptions{
+		Namespace: jobs[0].JobSummary.Namespace,
+	}
+
+	job, _, err := jobClient.Info(jobs[0].ID, q)
+
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad job: %s", err.Error())
+	}
+
+	st.Update("Getting latest deployments for job")
+	deploy, _, err := jobClient.LatestDeployment(*job.ID, q)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch latest deployment for Nomad job: %s", err.Error())
+	} else if deploy == nil {
+		st.Update("No active deployment for Nomad job")
+		return err
+	}
+
+	canaryDeployment := false
+	groupsToPromote := make([]string, len(deploy.TaskGroups))
+	for taskGroupName, taskGroup := range deploy.TaskGroups {
+		if r.config.Groups != nil {
+			if isElementExist(r.config.Groups, taskGroupName) && taskGroup.DesiredCanaries != 0 {
+				canaryDeployment = true
+				groupsToPromote = append(groupsToPromote, taskGroupName)
+				continue
+			}
+		} else if taskGroup.DesiredCanaries != 0 {
+			canaryDeployment = true
+			// if no groups to promote are specified in the config, promote all groups
+			// that have canaries
+			groupsToPromote = append(groupsToPromote, taskGroupName)
+		}
+	}
+	if !canaryDeployment {
+		log.Info("Canaries not detected")
+		return nil
+	}
+
+	// check each task group to be promoted; if the canary allocs aren't healthy,
+	//   check again in 5 seconds
+	// TODO: Force timeout if exceeds healthy deadline or progress deadline of job
+	var currentTaskGroupState *api.DeploymentState
+	var groupHealthy bool
+	for _, group := range groupsToPromote {
+		if group != "" {
+			currentTaskGroupState = deploy.TaskGroups[group]
+			groupHealthy = false
+			for !groupHealthy {
+				if currentTaskGroupState.HealthyAllocs < len(currentTaskGroupState.PlacedCanaries) {
+					time.Sleep(5 * time.Second)
```

Comment:
The top-level function this is in `resourceJobCreate` has a `context.Context`. Should we be using that here rather than having a fixed sleep without a way of breaking it? Is there any kind of backoff we should be using?

Reply:
@tgross took some time, but I think I've got this covered! 

https://github.com/hashicorp/waypoint/pull/2938/files#diff-8465d427b10de27ab3337b90735cb0728c5f62a8ce0612476219fbb6aacc2574R148-R176

I have currently arbitrarily set the context deadline to 5 minutes here ([since it's the default](https://www.nomadproject.io/docs/job-specification/update#healthy_deadline)). Either in this PR or as an enhancement later I want to update this to pull the `HealthyDeadline` from the task group currently being assessed, and possibly also use child Go routines to concurrently check the task groups. Then if one of the task groups doesn't meet its healthy deadline before others don't, we may not have to wait extra time.